File: P2/Civitas/src/civitas/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author Salva
 */
public class TituloPropiedad {
    //atributos de instancia
    private float alquilerBase;
    private static final float factorInteresesHipoteca=1.1f;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    
    //atributo de referencia 
    private Jugador propietario;
    
    //metodos
    
    TituloPropiedad(String nombre, float alquBase, float factrevalor, float hipbase
            , float precioComp, float precioEdifi){
        nombre = nombre;
        alquilerBase = alquBase;
        factorRevalorizacion = factrevalor;
        hipotecaBase = hipbase;
        precioCompra = precioComp;
        precioEdificar = precioEdifi;
        propietario = null;
        numCasas = numHoteles = 0;
        hipotecado = false;
        
    }
    
    @Override
    public String toString(){
        String salida = "Nombre: " + nombre +
                        "\nAlquiler Base: " + alquilerBase + 
                        "\nFactor de Revalorización: " + factorRevalorizacion + 
                        "\nFactor Intereses Hipoteca: " + factorInteresesHipoteca +
                        "\nHipoteca Base: " + hipotecaBase + 
                        "\nHipotecado: " + hipotecado + 
                        "\nNúmero de casas: " + numCasas + 
                        "\nNúmero de Hoteles: " + numHoteles +
                        "\nPrecio de compra: " + precioCompra +
                        "\nPrecio de Edificar: " + precioEdificar;
        return salida;
                        
    }
    
    private float getPrecioAlquiler(){
      //  PrecioAlquiler=AlquilerBase*(1+(NumCasas*0.5)+(NumHoteles*2.5))
      float precio = alquilerBase*(1f+(numCasas*0.5f) + (numHoteles*2.5f));
      if (hipotecado || propietarioEncarcelado())
          precio = 0;
      return precio;
    }
        
    private float getImporteHipoteca(){
        //CantidadRecibida=HipotecaBase*(1+(numCasas*0.5)+(numHoteles*2.5))
        float importe = hipotecaBase*(1+(numCasas*0.5f) + (numHoteles*2.5f));
        return importe; 
    }
    
    float getImporteCancelarHipoteca(){
        float importe = getImporteHipoteca() * factorInteresesHipoteca;
        return importe;
    }
    
    void actualizaPropietarioPorConversion(Jugador jugador){
        propietario = jugador;
    }
    
    boolean cancelarHipoteca(Jugador jugador){
        
    }
    
    int cantidadCasasHoteles(){
        return getNumCasas() + getNumHoteles();
    }
    
    boolean comprar(Jugador jugador){
        
    }
    
    boolean construirCasa(Jugador jugador){
        
    }
    
    boolean construirHotel(Jugador jugador){
        
    }
    
    boolean derruirCasas(int n, Jugador jugador){
        if (esEsteElPropietario(jugador) && getNumCasas() >= n){
            numCasas -= n;
            return true;
        }
        return false;
    }
    
    private boolean esEsteElPropietario(Jugador jugador){
        return jugador == getPropietario();
    }
    
    public boolean getHipotecado(){
        return hipotecado;
    }
    
    
    String getNombre(){
        return nombre;
    }
    
    int getNumCasas(){
        return numCasas;
    }
    
    int getNumHoteles(){
        return numHoteles;
    }
    
    
    float getPrecioCompra(){
        return precioCompra;
    }
    
    float getPrecioEdificar(){
        return precioEdificar;
    }
    
    private float getPrecioVenta(){
        float precio = getPrecioCompra() + getPrecioEdificar()*cantidadCasasHoteles()*factorRevalorizacion;
    }
    
    Jugador getPropietario(){
        return propietario;
    }
    
    boolean hipotecar(Jugador jugador){
        
    }
    
    private boolean propietarioEncarcelado(){
        return getPropietario().isEncarcelado(); 
    }
    
    boolean tienePropietario(){
        return propietario != null;
    }
    
    void tramitarAlquiler(Jugador jugador){
       if (tienePropietario() && !esEsteElPropietario(jugador)){
           jugador.pagaAlquiler(getPrecioAlquiler());
           getPropietario().recibe(getPrecioAlquiler());
       }
           
    }
    
    boolean vender(Jugador jugador){
        if (esEsteElPropietario(jugador) && !hipotecado){
            jugador.recibe(getPrecioVenta());
            propietario = null;
            numCasas= numHoteles = 0;
        }
        else{
            return false;
        }
    }
    
    
}

File: P3/civitasruby/lib/civitas/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
# Salva Romero
module Civitas
  class Titulo_propiedad
    attr_reader :nombre, :precio_compra, :alquiler_base, :factor_revalorizacion,
      :hipoteca_base, :precio_edificar, :num_hoteles, :num_casas
    attr_accessor :hipotecada, :propietario

    def initialize(nom, precioComp, alquiler, factor, hipoteca, precioEdif)
      @nombre = nom
      @precio_compra = precioComp
      @alquiler_base = alquiler
      @factor_revalorizacion = factor
      @hipoteca_base = hipoteca
      @precio_edificar = precioEdif
      @hipotecada = false
      @num_hoteles = 0
      @num_casas = 0
      @propietario = nill
    end

    @override
    def to_s()
      "\n-TituloPropiedad: \nNombre: #{@nombre} \nHipotecada: #{@hipotecada}
         \nPrecio compra: #{@precioCompra} \nAlquiler base: #{@alquilerBase}
         \nFactor revalorizacion: #{@factorRevalorizacion} \nHipoteca base: #{@hipotecaBase}
         \nPrecio edificar: #{@precioEdificar} \nNúmero de hoteles: #{@numHoteles}
         \nNúmero de casas: #{@numCasas}\n\n"
    end

    def get_precio_alquiler
      res = @alquiler_base * (1+(@num_casas*0.5)
        + (@num_hoteles*2.5))
      if hipotecado == true or propietario_encarcelado
        res = 0
      end
      return res
    end

    def get_importe_hipoteca
      (@hipoteca_base*(1+(@num_casas*0.5) + (@num_hoteles*2.5)))
    end

    def get_importe_cancelar_hipoteca
      (get_importe_hipoteca * @@factor_intereses_hipoteca)
    end

    def actualiza_propietario_por_conversion(jugador)
      @propietario = jugador
    end

    def cancelar_hipoteca

    end

    def cantidad_casas_hoteles
      (get_num_casas + get_num_hoteles)
    end

    def comprar(jugador)

    end

    def construir_casa(jugador)

    end

    def construir_hotel(jugador)

    end

    def es_este_el_propietario(jugador)
      res = false
      if jugador == get_propietario
        res = true
      end
      return res
    end

    def get_hipotecado
      @hipotecado
    end

    def get_nombre
      @nombre
    end

    def get_casas
      @num_casas
    end

    def num_hoteles
      @num_hoteles
    end

    def get_precio_edificar
      @precio_edificar
    end

    def get_precio_venta
      (get_precio_compra + get_precio_edificar*
        cantidad_casas_hoteles*@factor_revalorizacion)
    end

    def get_propietario
      @propietario
    end

    def hipotecar(jugador)

    end

    def tramitar_alquiler(jugador)
      if tiene_propietario == true and es_este_el_propietario(jugador) == false
        jugador.paga_alquiler(get_precio_alquiler)
        @propietario.recibe(get_precio_alquiler)
      end
    end

    def propietario_encarcelado
      res = false
      if @propietario.encarcelado == true
        res = true
      end
      return res
    end

    def cantidad_casas_hoteles
      (@num_casas + @num_hoteles)
    end

    def derruir_casas(n, jugador)
      res = false
      if es_este_el_propietario(jugador) == true and @num_casas >= n
        @num_casas = n
        res = true
      end
      return res
    end

    def vender(jugador)
      res = false
      if(@hipotecado == false and es_este_el_propietario(jugador) == true)
        jugador.recibe(get_precio_venta())
        derruir_casas(@num_casas,jugador)
        derruir_hoteles(@num_hoteles,jugador)
        actualiza_propietario_por_conversion(nil)
        res = true
      end
      return res
    end


  end
end
